### PR TITLE
feat: add support for customizing root components

### PR DESCRIPTION
**Pull request by aam229**
This PR adds support for customizing the rendering of the root components. It works by defining a rootComponent path in the config file which points to a file that defines functions for rendering the server/client root components. I think it takes care of bdefore/universal-redux#48 but I'm not familiar enough with Relay/Falcor to be sure. 

For now the rootComponent.js included in this PR does not support any form of async loading. It would be easy to update it to use [Rezonans/redux-async-connect](https://github.com/Rezonans/redux-async-connect) if you want to support async loading by default. Let me know if that's something you would like me to do.

If you are curious, here is a [gist](https://gist.github.com/aam229/2431c99fcad1f2e291c2) for a custom async loading solution that fits my need but I'm not suggesting we use it by default. 

Please let me know what you think!

### CODE REVIEW

**Comment by bdefore** [comment]
This looks interesting @aam229, I'll have a look later today. In the meantime, can you fix the linting errors? Running `npm run lint` locally will identify them.

**Comment by aam229** [comment]
I apologize for that, it's fixed. Let me know if anything else needs to be fixed.

**Comment by bdefore** [comment]
@aam229 Well done on this, it fits nicely in with everything else and opens up some nice options for later. Before merging, I think it's worth keeping redux-async-connect in there by default, as otherwise there's no out-of-the-box way for UR to do serverside renders with data. Fortunately, with the `rootComponent` option, developers have a means to replace that with whatever they prefer. I'll be looking further into this today. Feel free to hit me up on Discord.

File: src/client.js [hunk 1]
```diff
@@ -11,29 +9,31 @@ import { render as renderDevtools } from './client/devtools';
 // as assigned in merge-configs.js
 import getRoutes from 'routes';
 import middleware from 'middleware';
+import {createForClient as createRootComponentForClient } from 'rootComponent';
```
Comment (sompylasar) [inline src/client.js]: Missing space: `{ createForClient`

File: src/client.js [hunk 2]
```diff
@@ -11,29 +9,31 @@ import { render as renderDevtools } from './client/devtools';
...
-}
-
-renderDevtools(component, store, dest);
+const store = createStore(middleware, history, window.__data);
+const routes = getRoutes(store);
+const devComponent = renderDevtools();
+
+// There is probably no need to be asynchronous here
+createRootComponentForClient(store, {routes, history})
+  .then(({root}) => {
```
Comment (sompylasar) [inline src/client.js]: Missing spaces: `({ root })`

File: src/client.js [hunk 3]
```diff
@@ -11,29 +9,31 @@ import { render as renderDevtools } from './client/devtools';
...
+    if (process.env.NODE_ENV !== 'production') {
+      window.React = React; // enable debugger
+      if (!dest || !dest.firstChild || !dest.firstChild.attributes || !dest.firstChild.attributes['data-react-checksum']) {
+        throw new Error('Server-side React render was discarded. Make sure that your initial render does not contain any client-side code.');
+      }
+    }
+
+    return devComponent ? createRootComponentForClient(store, {routes, history, devComponent}) : {};
+  })
+  .then(({root}) => {
```
Comment (sompylasar) [inline src/client.js]: Missing spaces: `({ root })`

File: src/helpers/rootComponent.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+export function createForServer(store, renderProps) {
+  const root = (
+    <Provider store={store} key="provider">
+      <div>
+        <RouterContext {...renderProps} />
+      </div>
+    </Provider>
+  );
+  return Promise.resolve({root});
```
Comment (sompylasar) [inline src/helpers/rootComponent.js]: Missing spaces: `({ root })`

File: src/helpers/rootComponent.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+  const root = (
+    <Provider store={store} key="provider">
+      <div>
+        {component}
+        {devComponent}
+      </div>
+    </Provider>
+  );
+
+  return Promise.resolve({root});
```
Comment (sompylasar) [inline src/helpers/rootComponent.js]: Missing spaces: `({ root })`

File: src/helpers/rootComponent.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    <Provider store={store} key="provider">
+      <div>
+        <RouterContext {...renderProps} />
+      </div>
+    </Provider>
+  );
+  return Promise.resolve({root});
+}
+
+export function createForClient(store, {routes, history, devComponent}) {
```
Comment (sompylasar) [inline src/helpers/rootComponent.js]: Missing spaces: `{ routes, history, devComponent }`

File: src/server/renderer.js [hunk 7]
```diff
@@ -44,15 +42,9 @@ export default (projectConfig, projectToolsConfig) => {
...
-            .then(() => {
-              const component = (
-                <Provider store={store} key="provider">
-                  <ReduxAsyncConnect {...renderProps} />
-                </Provider>
-              );
-
-              const content = html(config, tools.assets(), store, res._headers, component);
+          rootComponent.createForServer(store, renderProps)
+            .then(({root}) => {
```
Comment (sompylasar) [inline src/server/renderer.js]: Missing spaces: `({ root })`

**Comment by bdefore** [comment]
@sompylasar are you aware of a linting configuration to check for spaces around objects like that?

**Comment by sompylasar** [comment]
@bdefore Looks like this one: http://eslint.org/docs/rules/object-curly-spacing.html and this one as a companion for array literals: http://eslint.org/docs/rules/array-bracket-spacing

P.S. I use [eslint-config-airbnb](https://www.npmjs.com/package/eslint-config-airbnb) as a base config for coding style.

**Comment by bdefore** [comment]
@aam229 for redux-async-connect support, how do you feel about cherry-picking this in? https://github.com/bdefore/universal-redux/commit/232317d9751d77eabf56b3062e483181e26a1098 you should be able to use your custom rootComponent to override it.

**Comment by aam229** [comment]
Thank you for the feedback @sompylasar, I will fix the spacing issues. @bdefore, I will update the default `rootComponent.js` to use redux-async-connect. I think it will just be a copy and paste of some of the code in the link you provided. I'm at work so I will take care of these issues later today.

**Comment by bdefore** [comment]
Along with the redux-async-connect addition, I've added the eslint configuration for objects and array spacing along with the requisite fixes and merged the latest from master into my `custom-root-component` branch. @aam229 you should be able to cleanly merge into your own branch.

**Comment by aam229** [comment]
@bdefore great, thank you. Would you like to discuss taking this one step further to have a global hook system instead of only allowing the customization the root component rendering?

**Comment by bdefore** [comment]
@aam229 your head is exactly where mine is. allow projects to specify what router they want, or serverside data loading library. but I think it's best to pursue those in new PRs. i'm :+1: for merge if you're ok with the changes i have in my branch.

**Comment by aam229** [comment]
@bdefore  go for it.

**Comment by bdefore** [comment]
@aam229 i'm not authorized to make that change to your branch and would like your PR to be the one that's merged in

**Comment by aam229** [comment]
@bdefore I merged your changes into my branch. I'm pretty new to this whole process, is that all I need to do?

**Comment by bdefore** [comment]
@aam229 yup, we're good to go now. thanks!

**Comment by bdefore** [comment]
Released with 3.0.0-rc20

**Pull request merged**